File: apps/api/app/schemas/inventory.py
"""Pydantic schemas for inventory endpoints."""

from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


# ---------------------------------------------------------------------------
# Daily Counts
# ---------------------------------------------------------------------------


class CountItem(BaseModel):
    variety_id: UUID
    count_value: int | None = None
    is_done: bool = False


class CountSaveRequest(BaseModel):
    product_type_id: UUID
    count_date: date
    entered_by: str | None = None
    counts: list[CountItem]


class CountVarietyResponse(BaseModel):
    variety_id: str
    variety_name: str
    count_value: int | None
    is_done: bool
    entered_by: str | None
    updated_at: datetime | None = None


class CountProductLineResponse(BaseModel):
    product_line_id: str
    product_line_name: str
    varieties: list[CountVarietyResponse]


class CountSheetResponse(BaseModel):
    count_date: date
    product_type_id: str
    product_type_name: str
    sheet_complete: bool
    completed_by: str | None = None
    completed_at: datetime | None = None
    product_lines: list[CountProductLineResponse]


class CountSaveResponse(BaseModel):
    saved_count: int


class RecentCountItem(BaseModel):
    count_date: date
    count_value: int | None


# ---------------------------------------------------------------------------
# Customer Counts
# ---------------------------------------------------------------------------


class CustomerCountItem(BaseModel):
    variety_id: UUID
    customer_id: UUID
    bunch_size: int
    sleeve_type: str
    bunch_count: int | None = None
    is_done: bool = False


class CustomerCountSaveRequest(BaseModel):
    product_type_id: UUID
    count_date: date
    entered_by: str | None = None
    counts: list[CustomerCountItem]


class TemplateColumn(BaseModel):
    customer_id: str
    customer_name: str
    bunch_size: int
    sleeve_type: str


class CustomerCountVarietyResponse(BaseModel):
    variety_id: str
    variety_name: str
    is_done: bool
    counts: dict[str, int | None]


class CustomerCountProductLineResponse(BaseModel):
    product_line_id: str
    product_line_name: str
    varieties: list[CustomerCountVarietyResponse]
    totals: dict[str, int]


class CustomerCountSheetResponse(BaseModel):
    count_date: date
    product_type_id: str
    product_type_name: str
    sheet_complete: bool
    template_columns: list[TemplateColumn]
    product_lines: list[CustomerCountProductLineResponse]
    grand_totals: dict[str, int]


class CustomerCountSaveResponse(BaseModel):
    saved_count: int


# ---------------------------------------------------------------------------
# Estimates
# ---------------------------------------------------------------------------


class EstimateItem(BaseModel):
    variety_id: UUID
    pull_day: date
    estimate_value: int | None = None
    is_done: bool = False


class EstimateSaveRequest(BaseModel):
    product_type_id: UUID
    week_start: date
    entered_by: str | None = None
    estimates: list[EstimateItem]


class EstimateVarietyResponse(BaseModel):
    variety_id: str
    variety_name: str
    estimates: dict[str, int | None]
    is_done: bool


class EstimateProductLineResponse(BaseModel):
    product_line_id: str
    product_line_name: str
    varieties: list[EstimateVarietyResponse]


class EstimateSheetResponse(BaseModel):
    week_start: date
    product_type_id: str
    product_type_name: str
    sheet_complete: bool
    pull_days: list[date]
    last_week_actuals: dict[str, dict[str, int | None]]
    product_lines: list[EstimateProductLineResponse]


class EstimateSaveResponse(BaseModel):
    saved_count: int


# ---------------------------------------------------------------------------
# Comparison
# ---------------------------------------------------------------------------


class ComparisonDayData(BaseModel):
    estimate: int | None
    actual: int | None
    variance: int | None


class ComparisonVarietyResponse(BaseModel):
    variety_name: str
    days: dict[str, ComparisonDayData]


class ComparisonProductLineResponse(BaseModel):
    product_line_name: str
    varieties: list[ComparisonVarietyResponse]


class ComparisonSummary(BaseModel):
    total_estimated: int
    total_actual: int
    variance_pct: float | None


class ComparisonResponse(BaseModel):
    week_start: date
    pull_days: list[date]
    product_lines: list[ComparisonProductLineResponse]
    summary: ComparisonSummary


# ---------------------------------------------------------------------------
# Availability
# ---------------------------------------------------------------------------


class AvailabilityVariety(BaseModel):
    variety_name: str
    remaining_count: int | None
    estimate: int | None


class AvailabilityProductLine(BaseModel):
    product_line_name: str
    varieties: list[AvailabilityVariety]


class AvailabilityProductType(BaseModel):
    product_type_id: str
    product_type_name: str
    data_source: str
    counts_completed_at: datetime | None = None
    counts_completed_by: str | None = None
    product_lines: list[AvailabilityProductLine]


class AvailabilityResponse(BaseModel):
    date: date
    product_types: list[AvailabilityProductType]


# ---------------------------------------------------------------------------
# Sheet Completion
# ---------------------------------------------------------------------------


class SheetCompleteRequest(BaseModel):
    product_type_id: UUID
    sheet_type: str
    sheet_date: date
    completed_by: str | None = None


class SheetCompleteResponse(BaseModel):
    is_complete: bool
    completed_by: str | None = None
    completed_at: datetime | None = None


# ---------------------------------------------------------------------------
# Harvest Status
# ---------------------------------------------------------------------------


class HarvestStatusItem(BaseModel):
    variety_id: str
    variety_name: str
    product_line_name: str
    in_harvest: bool
    stems_per_bunch: int


class HarvestStatusUpdateItem(BaseModel):
    variety_id: UUID
    in_harvest: bool


class HarvestStatusUpdateRequest(BaseModel):
    updates: list[HarvestStatusUpdateItem]


class HarvestStatusUpdateResponse(BaseModel):
    updated_count: int


# ---------------------------------------------------------------------------
# Count Sheet Templates
# ---------------------------------------------------------------------------


class TemplateColumnInput(BaseModel):
    customer_id: UUID
    bunch_size: int
    sleeve_type: str


class TemplateUpdateRequest(BaseModel):
    columns: list[TemplateColumnInput]


class TemplateResponse(BaseModel):
    product_type_id: str
    columns: list[TemplateColumn]


# ---------------------------------------------------------------------------
# Pull Day Schedules
# ---------------------------------------------------------------------------


class PullDayUpdateRequest(BaseModel):
    week_start: date | None = None
    pull_days: list[int]


class PullDayScheduleResponse(BaseModel):
    week_start: date | None
    pull_days: list[int]
    pull_dates: list[date]
    is_default: bool

File: apps/web/src/types/inventory.ts
// ── Inventory Management Types ──────────────────────────────
// Matches API contract: specs/005-inventory-management/contracts/api-v1.md

// ── Daily Counts ────────────────────────────────────────────

export interface DailyCountVariety {
  variety_id: string;
  variety_name: string;
  count_value: number | null;
  is_done: boolean;
  entered_by: string | null;
  updated_at: string | null;
}

export interface DailyCountProductLine {
  product_line_id: string;
  product_line_name: string;
  varieties: DailyCountVariety[];
}

export interface DailyCountResponse {
  count_date: string;
  product_type_id: string;
  product_type_name: string;
  sheet_complete: boolean;
  completed_by: string | null;
  completed_at: string | null;
  product_lines: DailyCountProductLine[];
}

export interface CountEntry {
  variety_id: string;
  count_value: number | null;
  is_done: boolean;
}

export interface CountSaveRequest {
  product_type_id: string;
  count_date: string;
  entered_by: string;
  counts: CountEntry[];
}

export interface CountSaveResponse {
  saved_count: number;
}

export interface RecentCount {
  count_date: string;
  count_value: number | null;
}

// ── Customer Counts ─────────────────────────────────────────

export interface TemplateColumn {
  customer_id: string;
  customer_name: string;
  bunch_size: number;
  sleeve_type: string;
}

export interface CustomerCountVariety {
  variety_id: string;
  variety_name: string;
  is_done: boolean;
  counts: Record<string, number | null>;
}

export interface CustomerCountTotals {
  [key: string]: number;
}

export interface CustomerCountGrandTotals extends CustomerCountTotals {
  total_customer_bunched: number;
  total_remaining: number;
  total_all_bunched: number;
}

export interface CustomerCountProductLine {
  product_line_id: string;
  product_line_name: string;
  varieties: CustomerCountVariety[];
  totals: CustomerCountTotals;
}

export interface CustomerCountResponse {
  count_date: string;
  product_type_id: string;
  product_type_name: string;
  sheet_complete: boolean;
  template_columns: TemplateColumn[];
  product_lines: CustomerCountProductLine[];
  grand_totals: CustomerCountGrandTotals;
}

export interface CustomerCountEntry {
  variety_id: string;
  customer_id: string;
  bunch_size: number;
  sleeve_type: string;
  bunch_count: number | null;
  is_done: boolean;
}

export interface CustomerCountSaveRequest {
  product_type_id: string;
  count_date: string;
  entered_by: string;
  counts: CustomerCountEntry[];
}

// ── Estimates ───────────────────────────────────────────────

export interface EstimateVariety {
  variety_id: string;
  variety_name: string;
  estimates: Record<string, number | null>;
  is_done: boolean;
}

export interface EstimateProductLine {
  product_line_id: string;
  product_line_name: string;
  varieties: EstimateVariety[];
}

export interface EstimateResponse {
  week_start: string;
  product_type_id: string;
  product_type_name: string;
  sheet_complete: boolean;
  pull_days: string[];
  last_week_actuals: Record<string, Record<string, number>>;
  product_lines: EstimateProductLine[];
}

export interface EstimateEntry {
  variety_id: string;
  pull_day: string;
  estimate_value: number | null;
  is_done: boolean;
}

export interface EstimateSaveRequest {
  product_type_id: string;
  week_start: string;
  entered_by: string;
  estimates: EstimateEntry[];
}

// ── Comparison ──────────────────────────────────────────────

export interface ComparisonDay {
  estimate: number | null;
  actual: number | null;
  variance: number | null;
}

export interface ComparisonVariety {
  variety_name: string;
  days: Record<string, ComparisonDay>;
}

export interface ComparisonProductLine {
  product_line_name: string;
  varieties: ComparisonVariety[];
}

export interface ComparisonSummary {
  total_estimated: number;
  total_actual: number;
  variance_pct: number | null;
}

export interface ComparisonResponse {
  week_start: string;
  pull_days: string[];
  product_lines: ComparisonProductLine[];
  summary: ComparisonSummary;
}

// ── Availability ────────────────────────────────────────────

export interface AvailabilityVariety {
  variety_name: string;
  color_hex: string | null;
  remaining_count: number | null;
  estimate: number | null;
}

export interface AvailabilityProductLine {
  product_line_name: string;
  varieties: AvailabilityVariety[];
}

export interface AvailabilityProductType {
  product_type_id: string;
  product_type_name: string;
  data_source: "actual_counts" | "estimates_only";
  counts_completed_at: string | null;
  counts_completed_by: string | null;
  product_lines: AvailabilityProductLine[];
}

export interface AvailabilityResponse {
  date: string;
  product_types: AvailabilityProductType[];
}

// ── Sheet Completion ────────────────────────────────────────

export type SheetType = "daily_count" | "customer_count" | "estimate";

export interface SheetCompleteRequest {
  product_type_id: string;
  sheet_type: SheetType;
  sheet_date: string;
  completed_by: string;
}

export interface SheetCompleteResponse {
  is_complete: boolean;
  completed_by: string;
  completed_at: string;
}

export interface SheetUncompleteResponse {
  is_complete: boolean;
}

// ── Harvest Status ──────────────────────────────────────────

export interface HarvestStatusEntry {
  variety_id: string;
  variety_name: string;
  product_line_name: string;
  in_harvest: boolean;
  stems_per_bunch: number;
}

export interface HarvestStatusUpdate {
  variety_id: string;
  in_harvest: boolean;
}

export interface HarvestStatusUpdateRequest {
  updates: HarvestStatusUpdate[];
}

export interface HarvestStatusUpdateResponse {
  updated_count: number;
}

// ── Count Sheet Templates ───────────────────────────────────

export interface CountSheetTemplateResponse {
  product_type_id: string;
  columns: TemplateColumn[];
}

export interface TemplateColumnInput {
  customer_id: string;
  bunch_size: number;
  sleeve_type: string;
}

export interface CountSheetTemplateSaveRequest {
  columns: TemplateColumnInput[];
}

// ── Pull Day Schedule ───────────────────────────────────────

export interface PullDayScheduleResponse {
  week_start: string;
  pull_days: number[];
  pull_dates: string[];
  is_default: boolean;
}

export interface PullDayScheduleSaveRequest {
  week_start: string | null;
  pull_days: number[];
}

// ── Count Audit ─────────────────────────────────────────────

export interface CountAuditEntry {
  id: string;
  variety_id: string;
  variety_name: string;
  action: "set" | "add" | "remove";
  amount: number;
  resulting_total: number;
  entered_by: string;
  created_at: string;
}
